# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# Importación de librerías

In [128]:
import pandas as pd
import numpy as np
from numpy.random import RandomState

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

from scipy import stats as st

# Cargue y análisis de datos

In [129]:
df0 = pd.read_csv('/datasets/geo_data_0.csv')
df1 = pd.read_csv('/datasets/geo_data_1.csv')
df2 = pd.read_csv('/datasets/geo_data_2.csv')

In [130]:
df0.info()

print(df0.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647


In [131]:
df1.info()

print(df1.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305


In [132]:
df2.info()

print(df2.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


In [133]:
# identificación de valores nulos
print("Valores nulos en df0:")
print(df0.isna().sum())

print("\nValores nulos en df1:")
print(df1.isna().sum())

print("\nValores nulos en df2:")
print(df2.isna().sum())

Valores nulos en df0:
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Valores nulos en df1:
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Valores nulos en df2:
id         0
f0         0
f1         0
f2         0
product    0
dtype: int64


In [134]:
# identificación de valores duplicados
print("Valores duplicados en df0:", df0.duplicated().sum())

print("\nValores duplicados en df1:", df1.duplicated().sum())

print("\nValores duplicados en df2:", df2.duplicated().sum())


Valores duplicados en df0: 0

Valores duplicados en df1: 0

Valores duplicados en df2: 0


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - En esta sección dejaste una base sólida: importaste las librerías clave para el flujo del proyecto, cargaste las tres regiones de forma ordenada y verificaste estructura, tipos de datos, valores nulos y duplicados con resultados claros. Con esto, el notebook arranca con datos bien validados y listo para pasar a modelado y evaluación sin ruido por calidad de datos.
</div>


# Entrenamiento del modelo

In [135]:
# división del df0 en características y objetivo
features = df0[['f0', 'f1', 'f2']]
target = df0['product']

# división del conjunto de datos en entrenamiento y validación
features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=42)

In [136]:
# creación y entrenamiento del modelo
model = LinearRegression() # condición mencionada
model.fit(features_train, target_train)

# validación del modelo
target_pred = model.predict(features_valid)

In [137]:
# guardado de las predicciones y respuestas correctas
df_predictions = pd.DataFrame({'predictions': target_pred, 'answers': target_valid})
print(df_predictions.head())

       predictions     answers
75721   101.901017  122.073350
80184    78.217774   48.738540
19864   115.266901  131.338088
76699   105.618618   88.327757
92991    97.980185   36.959266


In [138]:
# cálculo del volumen medio de reservas
average_vol_reserves = df_predictions['predictions'].mean()
print(f"Volumen medio de las reservas predicho: {average_vol_reserves:.4f} miles de barriles")

# cálculo del RMSE
rmse = mean_squared_error(df_predictions['answers'], df_predictions['predictions']) ** 0.5

print(f"Raíz del error cuadrático medio (RMSE): {rmse:.4f} miles de barriles")

Volumen medio de las reservas predicho: 92.3988 miles de barriles
Raíz del error cuadrático medio (RMSE): 37.7566 miles de barriles


## Resultados
El volumen medio de las reservas predicho indica cuánto se espera que contenga un pozo promedio, mediante regresión lineal

El RMSE muestra el error promedio de las predicciones con respecto a las reservas reales. Cuanto más bajo sea el RMSE, mejor será el rendimiento del modelo

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La sección de entrenamiento está bien estructurada: separaste correctamente variables, aplicaste la proporción 75:25, utilizaste regresión lineal como exige el brief y calculaste tanto el promedio predicho como el RMSE con interpretación clara. La lógica del modelado es coherente y está alineada con los objetivos del proyecto.
</div>


# Creación de funciones

In [139]:
def entrenar_evaluar_modelo(data):
    # separación en características y objetivo
    features = data[['f0', 'f1', 'f2']]
    target = data['product']

    # división en conjunto de datos de entrenamiento y validación
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=42)

    # creación y entrenamiento del modelo
    model = LinearRegression()
    model.fit(features_train, target_train)

    # predicción del modelo
    target_pred = model.predict(features_valid)

    # cálculo del volumen medio de reservas y RMSE
    average_vol_reserves = np.mean(target_pred)
    rmse = mean_squared_error(target_valid, target_pred) ** 0.5

    return target_pred, target_valid, average_vol_reserves, rmse   

In [140]:
# aplicación de la función creada en todos los df

target_pred_0, target_valid_0, average_vol_0, rmse_0 = entrenar_evaluar_modelo(df0)
print(f"Región 0 - volumen medio de reservas {average_vol_0:.4f} miles de barriles, RMSE: {rmse_0:.4f} miles de barriles")

target_pred_1, target_valid_1, average_vol_1, rmse_1 = entrenar_evaluar_modelo(df1)
print(f"Región 1 - volumen medio de reservas {average_vol_1:.4f} miles de barriles, RMSE: {rmse_1:.4f} miles de barriles")

target_pred_2, target_valid_2, average_vol_2, rmse_2 = entrenar_evaluar_modelo(df2)
print(f"Región 2 - volumen medio de reservas {average_vol_2:.4f} miles de barriles, RMSE: {rmse_2:.4f} miles de barriles")

Región 0 - volumen medio de reservas 92.3988 miles de barriles, RMSE: 37.7566 miles de barriles
Región 1 - volumen medio de reservas 68.7129 miles de barriles, RMSE: 0.8903 miles de barriles
Región 2 - volumen medio de reservas 94.7710 miles de barriles, RMSE: 40.1459 miles de barriles


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La creación de la función mejora notablemente la estructura del proyecto y evita duplicación de código, cumpliendo con uno de los criterios clave de evaluación. Además, aplicaste el mismo procedimiento de forma consistente a las tres regiones, lo que permite una comparación clara y ordenada de resultados.
</div>


# Cálculo de ganancias

In [141]:
# datos mínimos para evitar pérdidas
total_cost = 100000000
wells = 200
min_cost_per_well = 500000
min_units = min_cost_per_well / 4500

# comparación de valores
def comparar_valores(average_vol, min_units):
    if average_vol >= min_units:
        print("La región puede evitar pérdidas según el volumen medio predicho")
    else:
        print("La región no alcanzará el mínimo para considerar rentable la inversión")

# impresión de los resultados
print(f"Cantidad mínima de reservas por pozo para ser rentable la inversión: {min_units:.4f} miles de barriles")
print(f"Volumen medio predicho en la Región 0: {average_vol_0:.4f} miles de barriles")
comparar_valores(average_vol_0, min_units)

print(f"\nVolumen medio predicho en la Región 1: {average_vol_1:.4f} miles de barriles")
comparar_valores(average_vol_1, min_units)

print(f"\nVolumen medio predicho en la Región 2: {average_vol_2:.4f} miles de barriles")
comparar_valores(average_vol_2, min_units)

Cantidad mínima de reservas por pozo para ser rentable la inversión: 111.1111 miles de barriles
Volumen medio predicho en la Región 0: 92.3988 miles de barriles
La región no alcanzará el mínimo para considerar rentable la inversión

Volumen medio predicho en la Región 1: 68.7129 miles de barriles
La región no alcanzará el mínimo para considerar rentable la inversión

Volumen medio predicho en la Región 2: 94.7710 miles de barriles
La región no alcanzará el mínimo para considerar rentable la inversión


## Conclusiones

La Región 2 presentó la mayor cantidad de volumen medio predicho, seguida de la Región 0 y por último la Región 1. Sin embargo, ninguna de las regiones logró superar el umbral mínimo de 111.111 miles de barriles para considerar la inversión como rentable.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - En esta sección integraste correctamente las condiciones del negocio al análisis técnico, traduciendo el presupuesto y los ingresos por unidad en un umbral claro de rentabilidad. La comparación entre el volumen medio predicho y el mínimo requerido muestra una conexión sólida entre modelo y toma de decisiones.
</div>


## Función para calcular la ganancia de un conjunto de pozos seleccionados

In [142]:
def calcular_ganancia_pozos_top(predictions):
    # valores predefinidos
    wells = 200
    price_per_unit=4500
    
    # ordenar las predicciones de mayor a menor
    sorted_predictions = np.sort(predictions)[::-1]

    # seleccionar las predicciones más altas (200 pozos)
    top_wells = sorted_predictions[:wells]

    # calcular el volumen total de reserva
    total_volume = top_wells.sum()

    # calcular la ganancia total
    total_profit = total_volume * price_per_unit

    return total_volume, total_profit   

In [143]:
# evaluación de la función
total_volume_0, total_profit_0 = calcular_ganancia_pozos_top(target_pred_0)
print(f"Región 0 - volumen total de reservas de los mejores 200 pozos: {total_volume_0:.4f} miles de barriles")
print(f"Región 0 - Ganancia total potencial: ${total_profit_0:.2f} USD")

total_volume_1, total_profit_1 = calcular_ganancia_pozos_top(target_pred_1)
print(f"\nRegión 1 - volumen total de reservas de los mejores 200 pozos: {total_volume_1:.4f} miles de barriles")
print(f"Región 1 - Ganancia total potencial: ${total_profit_1:.2f} USD")

total_volume_2, total_profit_2 = calcular_ganancia_pozos_top(target_pred_2)
print(f"\nRegión 2 - volumen total de reservas de los mejores 200 pozos: {total_volume_2:.4f} miles de barriles")
print(f"Región 2 - Ganancia total potencial: ${total_profit_2:.2f} USD")

Región 0 - volumen total de reservas de los mejores 200 pozos: 30881.4633 miles de barriles
Región 0 - Ganancia total potencial: $138966584.80 USD

Región 1 - volumen total de reservas de los mejores 200 pozos: 27748.7514 miles de barriles
Región 1 - Ganancia total potencial: $124869381.15 USD

Región 2 - volumen total de reservas de los mejores 200 pozos: 29728.8478 miles de barriles
Región 2 - Ganancia total potencial: $133779815.14 USD


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La función para seleccionar los 200 pozos con mayor predicción está bien planteada y refleja correctamente la lógica del negocio. La comparación del volumen total y la ganancia potencial por región permite avanzar con claridad hacia la decisión estratégica.
</div>


# Cálculo de los riesgos para cada región

In [144]:
def bootstrapping(predictions, n_samples=1000):
    state = np.random.RandomState(42)
    profits = []
    predictions = pd.Series(predictions)

    for i in range(n_samples):
        pred_sample = predictions.sample(n=200, replace=True, random_state=state)
                
        # ganancia total de la muestra
        profit = sum(pred_sample) * 4500 
        profits.append(profit)

    profits = pd.Series(profits)

    # cálculo del beneficio promedio promedio
    mean_profit = profits.mean()

    # cálculo del intervalo de confianza del 95%
    confidence_interval = st.t.interval(confidence=0.95, df=len(profits)-1, loc=profits.mean(), scale=profits.sem())

    # cálculo del riesgo de pérdidas
    lost_risk = np.mean(profits < 0) 

    return mean_profit, confidence_interval, lost_risk


In [145]:
profit_0, interval_0, risk_0 = bootstrapping(target_pred_0)
print(f"Región 0 - Beneficio promedio: ${profit_0:.4f}, Intervalo de confianza 95%: {interval_0}, Riesgo de pérdida: {risk_0:.1%}")

profit_1, interval_1, risk_1 = bootstrapping(target_pred_1)
print(f"\nRegión 1 - Beneficio promedio: ${profit_1:.4f}, Intervalo de confianza 95%: {interval_1}, Riesgo de pérdida: {risk_1:.1%}")

profit_2, interval_2, risk_2 = bootstrapping(target_pred_2)
print(f"\nRegión 2 - Beneficio promedio: ${profit_2:.4f}, Intervalo de confianza 95%: {interval_2}, Riesgo de pérdida: {risk_2:.1%}")

Región 0 - Beneficio promedio: $83233775.7367, Intervalo de confianza 95%: (83140954.37672688, 83326597.09663309), Riesgo de pérdida: 0.0%

Región 1 - Beneficio promedio: $61898748.1234, Intervalo de confianza 95%: (61713747.41670833, 62083748.83012258), Riesgo de pérdida: 0.0%

Región 2 - Beneficio promedio: $85317638.3266, Intervalo de confianza 95%: (85238495.40879843, 85396781.24441168), Riesgo de pérdida: 0.0%


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La sección de riesgos está bien encaminada: aplicaste bootstrapping con 1000 muestras, resumiste el beneficio promedio, calculaste un intervalo de confianza al 95% y reportaste el riesgo de pérdida de forma explícita. Con estos resultados ya tienes una base cuantitativa clara para sustentar la elección de región con criterio de riesgo.
</div>


# Conclusiones
Si comparamos los valores obtenidos por bootsrapping y por predicción de los 200 pozos más rentables en las tres regiones, tenemos diferencias importantes en los valores, a saber: <br>
Región 0: valor bootsrapping 83.23 millones USD VS valor 200 mejores pozos 138.96 millones USD <br>
Región 1: valor bootsrapping 61.89 millones USD VS valor 200 mejores pozos 124.86 millones USD <br>
Región 2: valor bootsrapping 85.31 millones USD VS valor 200 mejores pozos 133.77 millones USD <br>
Estos resultados tienen lógica ya que partimos del hecho que para la predicción se seleccionaron los 200 pozos con mejores rendimiento, algo estadísticamente difícil de lograr en la práctica.
Dado que la inversión para iniciar el proceso de extracción de petróleo sería de 100 millones USD, la opción más favorable para iniciar el proceso e intentar recuperar la inversión sería la Región número 2, aunque se debe tener en cuenta que dada la alta inversión.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - La conclusión conecta correctamente los resultados del top 200 con el escenario más realista del bootstrapping y fundamenta la elección en términos de beneficio esperado frente a la inversión. Se aprecia una interpretación coherente entre modelo, riesgo y decisión estratégica.
</div>


# Comentario General del Revisor

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a>  
    
¡Felicidades! Tu proyecto está **aprobado**. Has demostrado un manejo sólido del flujo completo de un proyecto de machine learning aplicado a un problema de negocio real, integrando modelado, análisis financiero y evaluación de riesgos de manera coherente.  

#### Puntos Positivos:

* **Procesamiento de datos:** Realizaste una validación adecuada de estructura, tipos de datos, valores nulos y duplicados en las tres regiones antes de modelar.
* **Visualización y análisis del rendimiento del modelo:** Calculaste correctamente el volumen medio predicho y el RMSE, interpretando su significado dentro del contexto del problema.
* **Ingeniería y organización del código:** Implementaste funciones para evitar duplicación y estructuraste el análisis de forma clara y replicable para cada región.
* **Comparación de escenarios y análisis de riesgo:** Aplicaste correctamente la técnica de bootstrapping con 1000 muestras, calculaste el beneficio promedio, el intervalo de confianza del 95% y el riesgo de pérdidas, y seleccionaste la región óptima bajo el criterio de riesgo < 2.5%.

Tu trabajo refleja comprensión técnica, lógica analítica y alineación con las condiciones del negocio. La recomendación final está bien fundamentada en métricas cuantitativas y análisis estadístico.

¡Estás desarrollando un criterio profesional sólido en ciencia de datos! 

</div>
